Extended log_plays callback plugin

lumatijev opened the pull request:
##### SUMMARY

Modified the log_plays callback module to extend its features and make it configurable.

When testing the callback before and after this commit (with default configuration) I did not find any major backward incompatibilities, except that I removed unnecessary white space at the end of each line in log files.

The idea behind this pull request is to make per host logging configurable (that was TODO in plugin's description) and to enable customization of the output that is written to the log files.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
log_plays

##### ANSIBLE VERSION
```
ansible 2.6.0
  config file = /home/lmatijevic/Development/ansible-test/ansible.cfg
  configured module search path = [u'/home/lmatijevic/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lmatijevic/.virtualenvs/ansible-test/lib/python2.7/site-packages/ansible
  executable location = /home/lmatijevic/.virtualenvs/ansible-test/bin/ansible
  python version = 2.7.14 (default, Dec 11 2017, 16:08:01) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

bcoca commented on lib/ansible/plugins/callback/log_plays.py:
```diff
@@ -9,23 +10,95 @@
     type: notification
     short_description: write playbook output to log file
-    version_added: historical
+    version_added: 2.6
```
the new options require the version_added, the plugin itself has been here before 2.6

lumatijev replied:
Fixed, thank you!

lumatijev commented:
I have some questions.

I added myself to the copyrights in the header comments. Is that ok?

Also, I would like to mention other authors in the description. Should I enlist all of the contributors to this plugin (3 of them)?

samdoran commented:
@lumatijev Adding yourself to the copyrights section is fine. You can add the other authors inside the `author` field. Not that this means they will be pinged for issues/PRs against this module.

lumatijev committed:
Extended log_plays callback plugin

piroux commented:
@lumatijev Now, I think that we will need to have `version_added: 2.7` :man_facepalming: :smile: 

@samdoran What is blocking to ship this PR?

bcoca commented on lib/ansible/plugins/callback/log_plays.py:
```diff
@@ -35,23 +109,34 @@
+        self.timestamp_format = self.get_option('timestamp_format')
+        self.data_format = self.get_option('data_format')
+
+        if not os.path.exists(self.log_directory):
```
it should also check if it can write there

lumatijev commented:
@piroux I was instructed to revert `version_added` by @bcoca:

> the new options require the version_added, the plugin itself has been here before 2.6

lumatijev committed:
Added check for log directory permissions

The log_plays callback plugin will now test if the configured log
directory exists and has required permissions (WX). Otherwise, the
plugin will be disabled and error message will be shown.

lumatijev commented:
bot_status

frisbee23 commented:
i seem to be missing the task name and/or module name in the log. i think it's a general problem, but i'm using this plugin ATM. would someone be so kind and explain to me why this -and propably many log/output plugins- seem to not print them?

lumatijev commented:
@frisbee23 I've just read your issue and as far as I can see it would not be hard to add option to log task names. And for module name I'm not sure at first, I'll have to look into this. As soon as I get some time I'll update this pull request to add this feature(s).

Nice catch, thank you!

lumatijev committed:
Added task and module names to log output

Resolves #42567

jonni83 commented:
I've been excited for a while to have this pull request merged, I'm hoping it'll be brought in soon to make this a viable plugin.  In particular, specifying the log location is necessary for my use.

While using this pull request, I receive the following error:
` [WARNING]: Failure using method (v2_runner_on_ok) in callback plugin
(<ansible.plugins.callback.log_plays.CallbackModule object at 0x2aaab34c0e90>): local variable 'task' referenced
before assignment
`
If you need any feature suggestions, i can.  Although I don't believe this thread is the appropriate place for them to live.

EpiqSty commented:
Hi,
@lumatijev, @bcoca , is there any chance to see it merged?
I just faced with the issue using such callback, it doesn't log anything from handlers.
But as it's not merged yet to not really sure, who could handle or at least give any advice for such kind of issues ;)

mdavranche commented:
Thanks for this great PR @lumatijev , that's definitely what was missing in Ansible.
Before adding a +1 to insist on the merge of this PR, maybe consider the following test.
That's my playbook:
```
---
- name: Test extended log plays
  hosts: clouddb
  strategy: free
  gather_facts: no
  tasks:

    - name: Sleep 1 sec
      command: /bin/sleep 1

    - name: Say hello
      debug:
        msg: Hello
```
That's what I have if I do not set the `timestamp_format` option:
```
Jun 01 2020 17:28:08 - Sleep 1 sec - command - OK - {"module_args": {"warn": true, "executable": null, "_uses_shell": false, "strip_empty_ends": true, "_raw_params": "/bin/sleep 1", "removes": null, "argv": null, "creates": null, "chdir": null, "stdin_add_newline": true, "stdin": null}} => {"stderr_lines": [], "cmd": ["/bin/sleep", "1"], "end": "2020-06-01 17:28:08.485540", "_ansible_no_log": false, "stdout": "", "changed": true, "rc": 0, "start": "2020-06-01 17:28:07.434385", "stderr": "", "delta": "0:00:01.051155", "stdout_lines": []}

Jun 01 2020 17:28:08 - Say hello - debug - OK - {"msg": "Hello", "changed": false, "_ansible_verbose_always": true, "_ansible_no_log": false}

```
(All good)
And that's what I have if I set it to `'%Y-%m-%d %H:%M:%S %Z'`:
```
2020-06-01 17:59:04 UTC - Say hello - debug - OK - {"module_args": {"warn": true, "executable": null, "_uses_shell": false, "strip_empty_ends": true, "_raw_params": "/bin/sleep 1", "removes": null, "argv": null, "creates": null, "chdir": null, "stdin_add_newline": true, "stdin": null}} => {"stderr_lines": [], "cmd": ["/bin/sleep", "1"], "end": "2020-06-01 17:59:03.767193", "_ansible_no_log": false, "stdout": "", "changed": true, "rc": 0, "start": "2020-06-01 17:59:02.712708", "stderr": "", "delta": "0:00:01.054485", "stdout_lines": []}

2020-06-01 17:59:04 UTC - Say hello - debug - OK - {"msg": "Hello", "changed": false, "_ansible_verbose_always": true, "_ansible_no_log": false}

```
The timestamp is OK, but the other information are not the wanted anymore (duplicate `Say hello`…).

mdavranche commented:
And here is the solution =)
https://github.com/ansible-collections/community.general/pull/442

jimi-c commented:
@mdavranche thanks! Since you've opened that issue in the proper collection repo, I'm going to go ahead and close this here now.

If there are any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

   * https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
   * https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code
   * IRC: #ansible or #ansible-devel on irc.freenode.net

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
